torch/csrc/jit/script/logging.h

```diff
@@ -0,0 +1,89 @@
+#pragma once
+
+#include <memory>
+#include <mutex>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include <torch/csrc/WindowsTorchApiMacro.h>
+
+namespace torch {
+namespace jit {
+namespace logging {
+
+class LoggerBase {
+ public:
+  TORCH_API virtual void addStatValue(
+      const std::string& stat_name,
+      int64_t val) = 0;
+  TORCH_API virtual int64_t getCounterValue(const std::string& name) const = 0;
+  virtual ~LoggerBase() {}
+};
+
+TORCH_API LoggerBase* getLogger();
+TORCH_API LoggerBase* setLogger(LoggerBase* logger);
+
+// No-op logger. This is the default and is meant to incur almost no runtime
+// overhead.
+
+class NoopLogger : public LoggerBase {
+ public:
+  void addStatValue(const std::string& stat_name, int64_t val) override {}
+  TORCH_API virtual int64_t getCounterValue(const std::string& name) const override {
+    return 0;
+  }
+  ~NoopLogger() {}
+};
+
+// Trivial locking logger. Pass in an instance of this to setLogger() to use it.
+// This keeps track of the sum of all statistics.
+//
+// NOTE: this is not written in a scalable way and should probably only be used
+// in the single-threaded case or for testing.
+class LockingLogger : public LoggerBase {
+ public:
+  TORCH_API void addStatValue(const std::string& stat_name, int64_t val) override;
+  TORCH_API virtual int64_t getCounterValue(const std::string& name) const override;
+  enum class AggregationType { SUM, AVG };
+  TORCH_API void setAggregationType(
+      const std::string& stat_name,
+      AggregationType type);
+  ~LockingLogger() {}
+
+ private:
+  mutable std::mutex m;
+  std::unordered_map<std::string, std::vector<int64_t>> raw_counters;
```

Comment:
why not just:
```
struct Counter {
   int64_t sum;
   size_t count;
};
std::unordered_map<std::string, Counter> raw_counters;
```
You get both sum an avg, and do not need to ever allocate memory.